R/src/feather-types.cpp

```diff
@@ -148,17 +148,25 @@ SEXP toSEXP(const PrimitiveArray* val) {
 
   case PrimitiveType::UTF8: {
     auto asChar = reinterpret_cast<const char*>(val->values);
+    auto total_bytes = meta.total_bytes;
     for (int i = 0; i < n; ++i) {
       uint32_t offset1 = val->offsets[i], offset2 = val->offsets[i + 1];
+      if(offset1 > total_bytes || offset2 > total_bytes) {
+        stop("Corrupt feather file: offsets out of bounds");
```

Comment:
This will only catch a very particular kind of file corruption (e.g. caused by the IO bug that was fixed in feather 0.3.1), namely that the file metadata was overwritten, or new data was accidentally appended to the file. Other kinds of file corruption would likely never make it this far (e.g. if the metadata itself was corrupt).